libhb/encavcodec.c

```diff
@@ -1173,6 +1274,50 @@ int encavcodecWork( hb_work_object_t * w, hb_buffer_t ** buf_in,
         return HB_WORK_DONE;
     }
 
+#if HB_PROJECT_FEATURE_QSV
+    // postponed encoder initialization, reused code from encavcodecInit()
+    if (hb_hwaccel_is_full_hardware_pipeline_enabled(pv->job) &&
+        hb_qsv_decode_is_enabled(pv->job) && pv->context->hw_frames_ctx == NULL && pv->job->qsv.ctx->hb_ffmpeg_qsv_hw_frames_ctx != NULL)
+    {
+        // use the same hw frames context as for decoder or filter graph hw frames context
+        pv->context->hw_frames_ctx = pv->job->qsv.ctx->hb_ffmpeg_qsv_hw_frames_ctx;
+        int open_ret = 0;
+        if ((open_ret = hb_avcodec_open(pv->context, pv->qsv_data.codec, &pv->qsv_data.av_opts, HB_FFMPEG_THREADS_AUTO)))
+        {
+            hb_log( "encavcodecWork: avcodec_open failed: %s", av_err2str(open_ret) );
+            return HB_WORK_ERROR;
+        }
+
+        /*
+        * Reload colorimetry settings in case custom
+        * values were set in the encoder_options string.
+        */
+        pv->job->color_prim_override     = pv->context->color_primaries;
```

Comment:
It's too late to set these, because the previous values have already been set to the mp4/mkv/webm container.

Reply:
Can you point our where?

I am exiting earlier from the init function https://github.com/galinart/HandBrake/blob/enable-ffmpeg-qsv-encoder/libhb/encavcodec.c#L947

Reply:
Somewhere in encavcodecInit, hopefully the avcodec_open won't change the context values.